Implementations/Algorithms/Pong_REINFORCE.py

```diff
@@ -53,41 +55,57 @@ def forward(self, state):
-                 baseline=None, r2g=True):
+    def __init__(self, env, config):
+        # Initialize attributes from parameters
         self.env = env
-        self.show_plot = show_plot
-        self.save_plot = save_plot
-
+        self.save_plot = config.save_plot
+        self.show_plot = config.show_plot
+        self.log_window_size = config.log_window_size
+        self.log_flush_freq = config.log_flush_freq
+        self.model_path = config.model_path
+        self.override_model = config.override_model
+
+        # Create network
         self.num_actions = self.env.action_space.n
-        self.policy_network = PolicyNetwork(WIDTH*HEIGHT, self.env.action_space.n)
+        self.policy_network = PolicyNetwork(FRAMES*WIDTH*HEIGHT, self.env.action_space.n)
 
-        self.torch_rand = torch_rand
-        self.baseline = baseline
-        self.r2g = r2g
+        self.torch_rand = config.torch_rand
+        self.baseline = config.baseline
+        self.r2g = config.reward_to_go
 
+        # Set network to compute using cuda
         print("Looking for GPU support...")
         using_cuda = torch.cuda.is_available()
         self.device = torch.device("cuda:0" if using_cuda else "cpu")
         print("using cuda:",using_cuda)
         self.policy_network.set_device(self.device)
 
-#         if cuda:
-#             self.policy_network.cuda()
-#             self.device = torch.device("cuda:0") # Uncomment this to run on GPU
-#         else:
-#             self.device = torch.device("cpu")
+        # Initializes the optimizer
+        self.optimizer = optim.Adam(self.policy_network.parameters(), lr=config.learning_rate)
+
+        # Loads model from file if it exists
+        if config.use_loaded_model and os.path.exists(self.model_path):
+            checkpoint = torch.load(self.model_path)
+            self.policy_network.load_state_dict(checkpoint['model_state_dict'])
+            self.optimizer.load_state_dict(checkpoint['optimizer_state_dict'])
+            self.policy_network.eval()
 
-        self.optimizer = optim.Adam(self.policy_network.parameters(), lr=learning_rate)
+    def save_model(self):
+        if not os.path.exists(os.path.dirname(self.model_path)) and self.override_model:
+            os.makedirs(os.path.dirname(self.model_path))
+            torch.save({
+                        'model_state_dict': self.policy_network.state_dict(),
+                        'optimizer_state_dict': self.optimizer.state_dict(),
+                        }, self.model_path)
 
     def get_action(self, state):
         if self.torch_rand:
-            logits = self.policy_network(state)
-            dist = Categorical(logits=logits)
+            probs = self.policy_network(state)
+            dist = Categorical(probs=probs)
             highest_prob_action = dist.sample()
             log_prob = dist.log_prob(highest_prob_action)
         else:
-            probs = self.policy_network(state).cpu()
+            probs = self.policy_network(state)
```

Comment:
If you run this on the GPU, won't it spit out an error and complain about needing to access it from the host memory?

Reply:
Done!